cranelift/codegen/src/isa/x64/abi.rs

```diff
@@ -732,12 +732,19 @@ impl ABIMachineSpec for X64ABIMachineSpec {
     }
 
     fn is_frame_setup_needed(
-        _is_leaf: bool,
-        _stack_args_size: u32,
-        _num_clobbered_callee_saves: usize,
-        _frame_storage_size: u32,
+        is_leaf: bool,
+        stack_args_size: u32,
+        num_clobbered_callee_saves: usize,
+        frame_storage_size: u32,
     ) -> bool {
-        true
+        (!is_leaf
+            // The function arguments that are passed on the stack are addressed
+            // relative to the Frame Pointer.
+            || stack_args_size > 0
+            || num_clobbered_callee_saves > 0
+            || frame_storage_size > 0)
+            // TODO
+            && !cfg!(target_os = "macos")
```

Comment:
Maybe my suggestion at https://github.com/bytecodealliance/wasmtime/pull/4469#discussion_r925815651 would work? "non-leaf" could be used by default and "always" could be used on macOS.